# Проверка гипотез по увеличению выручки в интернет-магазине —оценить результаты A/B теста

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Часть-1.-Приоритизация-гипотез." data-toc-modified-id="Часть-1.-Приоритизация-гипотез.-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Часть 1. Приоритизация гипотез.</a></span><ul class="toc-item"><li><span><a href="#Фреймворк-ICE" data-toc-modified-id="Фреймворк-ICE-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Фреймворк ICE</a></span></li><li><span><a href="#Фреймворк-RICE" data-toc-modified-id="Фреймворк-RICE-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Фреймворк RICE</a></span></li><li><span><a href="#Приоритизация-гипотез-при-применении-RICE-вместо-ICE" data-toc-modified-id="Приоритизация-гипотез-при-применении-RICE-вместо-ICE-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Приоритизация гипотез при применении RICE вместо ICE</a></span></li></ul></li><li><span><a href="#Часть-2.-Анализ-A/B-теста" data-toc-modified-id="Часть-2.-Анализ-A/B-теста-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Часть 2. Анализ A/B-теста</a></span><ul class="toc-item"><li><span><a href="#График-кумулятивной-выручки-по-группам." data-toc-modified-id="График-кумулятивной-выручки-по-группам.-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>График кумулятивной выручки по группам.</a></span></li><li><span><a href="#График-кумулятивного-среднего-чека-по-группам" data-toc-modified-id="График-кумулятивного-среднего-чека-по-группам-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>График кумулятивного среднего чека по группам</a></span></li><li><span><a href="#График-относительного-изменения-кумулятивного-среднего-чека-группы-B-к-группе-A" data-toc-modified-id="График-относительного-изменения-кумулятивного-среднего-чека-группы-B-к-группе-A-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>График относительного изменения кумулятивного среднего чека группы B к группе A</a></span></li><li><span><a href="#График-кумулятивной-конверсии-по-группам" data-toc-modified-id="График-кумулятивной-конверсии-по-группам-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>График кумулятивной конверсии по группам</a></span></li><li><span><a href="#График-относительного-изменения-кумулятивной-конверсии-группы-B-к-группе-A" data-toc-modified-id="График-относительного-изменения-кумулятивной-конверсии-группы-B-к-группе-A-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>График относительного изменения кумулятивной конверсии группы B к группе A</a></span></li><li><span><a href="#Точечный-график-количества-заказов-по-пользователям" data-toc-modified-id="Точечный-график-количества-заказов-по-пользователям-2.6"><span class="toc-item-num">2.6&nbsp;&nbsp;</span>Точечный график количества заказов по пользователям</a></span></li><li><span><a href="#95-й-и-99-й-перцентили-количества-заказов-на-пользователя" data-toc-modified-id="95-й-и-99-й-перцентили-количества-заказов-на-пользователя-2.7"><span class="toc-item-num">2.7&nbsp;&nbsp;</span>95-й и 99-й перцентили количества заказов на пользователя</a></span></li><li><span><a href="#Точечный-график-стоимостей-заказов" data-toc-modified-id="Точечный-график-стоимостей-заказов-2.8"><span class="toc-item-num">2.8&nbsp;&nbsp;</span>Точечный график стоимостей заказов</a></span></li><li><span><a href="#95-й-и-99-й-перцентили-стоимости-заказов" data-toc-modified-id="95-й-и-99-й-перцентили-стоимости-заказов-2.9"><span class="toc-item-num">2.9&nbsp;&nbsp;</span>95-й и 99-й перцентили стоимости заказов</a></span></li><li><span><a href="#Статистическую-значимость-различий-в-конверсии-между-группами-по-«сырым»-данным" data-toc-modified-id="Статистическую-значимость-различий-в-конверсии-между-группами-по-«сырым»-данным-2.10"><span class="toc-item-num">2.10&nbsp;&nbsp;</span>Статистическую значимость различий в конверсии между группами по «сырым» данным</a></span></li><li><span><a href="#Cтатистическая-значимость-различий-в-среднем-чеке-заказа-между-группами-по-«сырым»-данным" data-toc-modified-id="Cтатистическая-значимость-различий-в-среднем-чеке-заказа-между-группами-по-«сырым»-данным-2.11"><span class="toc-item-num">2.11&nbsp;&nbsp;</span>Cтатистическая значимость различий в среднем чеке заказа между группами по «сырым» данным</a></span></li><li><span><a href="#Статистическую-значимость-различий-в-конверсии-между-группами-по-«очищенным»-данным" data-toc-modified-id="Статистическую-значимость-различий-в-конверсии-между-группами-по-«очищенным»-данным-2.12"><span class="toc-item-num">2.12&nbsp;&nbsp;</span>Статистическую значимость различий в конверсии между группами по «очищенным» данным</a></span></li><li><span><a href="#Статистическую-значимость-различий-в-среднем-чеке-заказа-между-группами-по-«очищенным»-данным" data-toc-modified-id="Статистическую-значимость-различий-в-среднем-чеке-заказа-между-группами-по-«очищенным»-данным-2.13"><span class="toc-item-num">2.13&nbsp;&nbsp;</span>Статистическую значимость различий в среднем чеке заказа между группами по «очищенным» данным</a></span></li><li><span><a href="#Общий-вывод" data-toc-modified-id="Общий-вывод-2.14"><span class="toc-item-num">2.14&nbsp;&nbsp;</span>Общий вывод</a></span></li></ul></li></ul></div>

# Задание   

Вместе с отделом маркетинга подготовлен список гипотез для увеличения выручки.
Приоритизировать гипотезы, запустить A/B-тест и проанализировать результаты.

In [1]:
import warnings
warnings.filterwarnings('ignore')

import pandas as pd
import numpy as np
import scipy.stats as stats
import seaborn as sns
import matplotlib.pyplot as plt
from collections import defaultdict
import datetime as dt
from plotly.offline import download_plotlyjs, init_notebook_mode, plot, iplot
import plotly.graph_objs as go

init_notebook_mode(connected=True)
pd.set_option('display.max_colwidth', -1)

## Часть 1. Приоритизация гипотез.
В файле /datasets/hypothesis.csv 9 гипотез по увеличению выручки интернет-магазина с указанными параметрами Reach, Impact, Confidence, Effort.  

**Задача**
- Применить фреймворк ICE для приоритизации гипотез. Отсортировать их по убыванию приоритета.
- Применить фреймворк RICE для приоритизации гипотез. Отсортировать их по убыванию приоритета.
- Указать, как изменилась приоритизация гипотез при применении RICE вместо ICE. Объяснить, почему так произошло.  

**Описание данных**  

Файл /datasets/hypothesis.csv.
- Hypothesis — краткое описание гипотезы;  
- Reach — охват пользователей по 10-балльной шкале;  
- Impact — влияние на пользователей по 10-балльной шкале;  
- Confidence — уверенность в гипотезе по 10-балльной шкале;  
- Efforts — затраты ресурсов на проверку гипотезы по 10-балльной шкале.  
Чем больше значение Efforts, тем дороже проверка гипотезы.

In [2]:
# Открываем файл
hypothesis = pd.read_csv('/datasets/hypothesis.csv')  
hypothesis.columns = ['hypothesis', 'reach', 'impact', 'confidence', 'efforts']
hypothesis

hypothesis  \
0  Добавить два новых канала привлечения трафика, что позволит привлекать на 30% больше пользователей             
1  Запустить собственную службу доставки, что сократит срок доставки заказов                                      
2  Добавить блоки рекомендаций товаров на сайт интернет магазина, чтобы повысить конверсию и средний чек заказа   
3  Изменить структура категорий, что увеличит конверсию, т.к. пользователи быстрее найдут нужный товар            
4  Изменить цвет фона главной страницы, чтобы увеличить вовлеченность пользователей                               
5  Добавить страницу отзывов клиентов о магазине, что позволит увеличить количество заказов                       
6  Показать на главной странице баннеры с актуальными акциями и распродажами, чтобы увеличить конверсию           
7  Добавить форму подписки на все основные страницы, чтобы собрать базу клиентов для email-рассылок               
8  Запустить акцию, дающую скидку на товар в день рождения                                                        

   reach  impact  confidence  efforts  
0  3      10      8           6        
1  2      5       4           10       
2  8      3       7           3        
3  8      3       3           8        
4  3      1       1           1        
5  3      2       2           3        
6  5      3       8           3        
7  10     7       8           5        
8  1      9       9           5

### Фреймворк ICE
Применим фреймворк ICE для приоритизации гипотез. Отсортируем их по убыванию приоритета.

In [3]:
hypothesis['ICE'] = hypothesis['impact'] * hypothesis['confidence'] / hypothesis['efforts']
hypothesis_ICE = hypothesis[['hypothesis', 'ICE']].sort_values(by='ICE', ascending=False).round(2)
hypothesis_ICE

hypothesis  \
8  Запустить акцию, дающую скидку на товар в день рождения                                                        
0  Добавить два новых канала привлечения трафика, что позволит привлекать на 30% больше пользователей             
7  Добавить форму подписки на все основные страницы, чтобы собрать базу клиентов для email-рассылок               
6  Показать на главной странице баннеры с актуальными акциями и распродажами, чтобы увеличить конверсию           
2  Добавить блоки рекомендаций товаров на сайт интернет магазина, чтобы повысить конверсию и средний чек заказа   
1  Запустить собственную службу доставки, что сократит срок доставки заказов                                      
5  Добавить страницу отзывов клиентов о магазине, что позволит увеличить количество заказов                       
3  Изменить структура категорий, что увеличит конверсию, т.к. пользователи быстрее найдут нужный товар            
4  Изменить цвет фона главной страницы, чтобы увеличить вовлеченность пользователей                               

     ICE  
8  16.20  
0  13.33  
7  11.20  
6  8.00   
2  7.00   
1  2.00   
5  1.33   
3  1.12   
4  1.00

- **Приоритетная гипотиза: Запустить акцию, дающую скидку на товар в день рождения, ICE: 16.20**

### Фреймворк RICE
Применим фреймворк RICE для приоритизации гипотез. Отсортируем их по убыванию приоритета.

In [4]:
hypothesis['RICE'] = hypothesis['reach'] * hypothesis['impact'] * hypothesis['confidence'] / hypothesis['efforts']
hypothesis_RICE = hypothesis[['hypothesis', 'RICE']].sort_values(by='RICE', ascending=False)
hypothesis_RICE

hypothesis  \
7  Добавить форму подписки на все основные страницы, чтобы собрать базу клиентов для email-рассылок               
2  Добавить блоки рекомендаций товаров на сайт интернет магазина, чтобы повысить конверсию и средний чек заказа   
0  Добавить два новых канала привлечения трафика, что позволит привлекать на 30% больше пользователей             
6  Показать на главной странице баннеры с актуальными акциями и распродажами, чтобы увеличить конверсию           
8  Запустить акцию, дающую скидку на товар в день рождения                                                        
3  Изменить структура категорий, что увеличит конверсию, т.к. пользователи быстрее найдут нужный товар            
1  Запустить собственную службу доставки, что сократит срок доставки заказов                                      
5  Добавить страницу отзывов клиентов о магазине, что позволит увеличить количество заказов                       
4  Изменить цвет фона главной страницы, чтобы увеличить вовлеченность пользователей                               

    RICE  
7  112.0  
2  56.0   
0  40.0   
6  40.0   
8  16.2   
3  9.0    
1  4.0    
5  4.0    
4  3.0

- **Приоритетная гипотиза:  
Добавить форму подписки на все основные страницы, чтобы собрать базу клиентов для email-рассылок, RICE: 112.0**

### Приоритизация гипотез при применении RICE вместо ICE
Укажем, как изменилась приоритизация гипотез при применении RICE вместо ICE. Объясним, почему так произошло.

In [5]:
priority_of_hypothesis = hypothesis_ICE.merge(hypothesis_RICE, on='hypothesis')
priority_of_hypothesis

hypothesis  \
0  Запустить акцию, дающую скидку на товар в день рождения                                                        
1  Добавить два новых канала привлечения трафика, что позволит привлекать на 30% больше пользователей             
2  Добавить форму подписки на все основные страницы, чтобы собрать базу клиентов для email-рассылок               
3  Показать на главной странице баннеры с актуальными акциями и распродажами, чтобы увеличить конверсию           
4  Добавить блоки рекомендаций товаров на сайт интернет магазина, чтобы повысить конверсию и средний чек заказа   
5  Запустить собственную службу доставки, что сократит срок доставки заказов                                      
6  Добавить страницу отзывов клиентов о магазине, что позволит увеличить количество заказов                       
7  Изменить структура категорий, что увеличит конверсию, т.к. пользователи быстрее найдут нужный товар            
8  Изменить цвет фона главной страницы, чтобы увеличить вовлеченность пользователей                               

     ICE   RICE  
0  16.20  16.2   
1  13.33  40.0   
2  11.20  112.0  
3  8.00   40.0   
4  7.00   56.0   
5  2.00   4.0    
6  1.33   4.0    
7  1.12   9.0    
8  1.00   3.0

- **Параметр охвата (Reach), использующийся в методе RICE изменил всю приоритетность, полученную по методу ICE,  
только 6 и 4 гипотезы остались на прежних местах.**

## Часть 2. Анализ A/B-теста   

Проведем A/B-тест и получим результаты, которые описаны в файлах /datasets/orders.csv и /datasets/visitors.csv.  

**Задача**  
Проанализируйте A/B-тест:  
1 Постройте график кумулятивной выручки по группам. Сделайте выводы и предположения.  
2 Постройте график кумулятивного среднего чека по группам. Сделайте выводы и предположения.  
3 Постройте график относительного изменения кумулятивного среднего чека группы B к группе A. Сделайте выводы и предположения.  
4 Постройте график кумулятивной конверсии по группам. Сделайте выводы и предположения.  
5 Постройте график относительного изменения кумулятивной конверсии группы B к группе A. Сделайте выводы и предположения.  
6 Постройте точечный график количества заказов по пользователям. Сделайте выводы и предположения.  
7 Посчитайте 95-й и 99-й перцентили количества заказов на пользователя. Выберите границу для определения аномальных пользователей. 
8 Постройте точечный график стоимостей заказов. Сделайте выводы и предположения.  
9 Посчитайте 95-й и 99-й перцентили стоимости заказов. Выберите границу для определения аномальных заказов.  
10 Посчитайте статистическую значимость различий в конверсии между группами по «сырым» данным. Сделайте выводы и предположения.  
11 Посчитайте статистическую значимость различий в среднем чеке заказа между группами по «сырым» данным. Сделайте выводы и предположения.  
12 Посчитайте статистическую значимость различий в конверсии между группами по «очищенным» данным. Сделайте выводы и предположения.  
13 Посчитайте статистическую значимость различий в среднем чеке заказа между группами по «очищенным» данным. Сделайте выводы и предположения.  
14 Примите решение по результатам теста и объясните его. Варианты решений: 1. Остановить тест, зафиксировать победу одной из групп. 2. Остановить тест, зафиксировать отсутствие различий между группами. 3. Продолжить тест.  

**Описание данных**  

Файл /datasets/orders.csv.   
- transactionId — идентификатор заказа;  
- visitorId — идентификатор пользователя, совершившего заказ;  
- date — дата, когда был совершён заказ;  
- revenue — выручка заказа;  
- group — группа A/B-теста, в которую попал заказ.  

Файл /datasets/visitors.csv.  
- date — дата;  
- group — группа A/B-теста;  
- visitors — количество пользователей в указанную дату в указанной группе A/B-теста

# Предобработка данных

In [6]:
# Открываем файл
orders = pd.read_csv('/datasets/orders.csv') 
orders.info()
orders

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1197 entries, 0 to 1196
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   transactionId  1197 non-null   int64 
 1   visitorId      1197 non-null   int64 
 2   date           1197 non-null   object
 3   revenue        1197 non-null   int64 
 4   group          1197 non-null   object
dtypes: int64(3), object(2)
memory usage: 46.9+ KB


transactionId   visitorId        date  revenue group
0     3667963787     3312258926  2019-08-15  1650     B   
1     2804400009     3642806036  2019-08-15  730      B   
2     2961555356     4069496402  2019-08-15  400      A   
3     3797467345     1196621759  2019-08-15  9759     B   
4     2282983706     2322279887  2019-08-15  2308     B   
...          ...            ...         ...   ...    ..   
1192  2662137336     3733762160  2019-08-14  6490     B   
1193  2203539145     370388673   2019-08-14  3190     A   
1194  1807773912     573423106   2019-08-14  10550    A   
1195  1947021204     1614305549  2019-08-14  100      A   
1196  3936777065     2108080724  2019-08-15  202740   B   

[1197 rows x 5 columns]

- Данные в порядке

In [7]:
visitors = pd.read_csv('/datasets/visitors.csv') 
visitors.info()
visitors

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 62 entries, 0 to 61
Data columns (total 3 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   date      62 non-null     object
 1   group     62 non-null     object
 2   visitors  62 non-null     int64 
dtypes: int64(1), object(2)
memory usage: 1.6+ KB


date group  visitors
0   2019-08-01  A     719     
1   2019-08-02  A     619     
2   2019-08-03  A     507     
3   2019-08-04  A     717     
4   2019-08-05  A     756     
..         ... ..     ...     
57  2019-08-27  B     720     
58  2019-08-28  B     654     
59  2019-08-29  B     531     
60  2019-08-30  B     490     
61  2019-08-31  B     718     

[62 rows x 3 columns]

- Данные в порядке  

Проверим нет ли пользователей, которые оказались в обеих группах одновременно.

In [8]:
inter = np.intersect1d(orders.query('group == "A"')['visitorId'].unique(), orders.query('group == "B"')['visitorId'].unique())

In [9]:
orders_filt = orders[~orders['visitorId'].isin(inter)]
orders_filt

transactionId   visitorId        date  revenue group
0     3667963787     3312258926  2019-08-15  1650     B   
1     2804400009     3642806036  2019-08-15  730      B   
3     3797467345     1196621759  2019-08-15  9759     B   
4     2282983706     2322279887  2019-08-15  2308     B   
5     182168103      935554773   2019-08-15  2210     B   
...         ...            ...          ...   ...    ..   
1191  3592955527     608641596   2019-08-14  16490    B   
1192  2662137336     3733762160  2019-08-14  6490     B   
1193  2203539145     370388673   2019-08-14  3190     A   
1194  1807773912     573423106   2019-08-14  10550    A   
1196  3936777065     2108080724  2019-08-15  202740   B   

[1016 rows x 5 columns]

- Удалили 181 пользователя попавших в обе группы.

In [10]:
group_sizes = orders.groupby(['visitorId', 'group']).size().reset_index()
group_sizes.columns = ['visitorId', 'group', 'count']
group_sizes = group_sizes.query('count > 1').sort_values(by='visitorId')
group_sizes

visitorId group  count
45    199603092   A     4    
58    237748145   A     4    
63    249864742   B     3    
94    366673373   A     2    
104   393266494   B     2    
...         ...  ..    ..    
1045  4109358064  A     2    
1050  4134830387  B     2    
1065  4186807279  B     2    
1079  4256040402  A     7    
1080  4256040402  B     4    

[64 rows x 3 columns]

In [11]:
potentially_broken_users_list = list(zip(group_sizes['visitorId'], group_sizes['group']))
potentially_broken_users_dict = defaultdict(list)
for user, group in potentially_broken_users_list:
    potentially_broken_users_dict[user].append(group)

inappropriate_users = [user for user in potentially_broken_users_dict if len(potentially_broken_users_dict[user]) > 1]
inappropriate_orders = orders.query('visitorId in @inappropriate_users')
inappropriate_orders_grouped = (inappropriate_orders
                        .groupby(['visitorId', 'group', 'date'])
                        .agg({'transactionId': 'nunique', 'revenue': 'sum'})
                   )
inappropriate_orders_grouped

transactionId  revenue
visitorId  group date                              
1230306981 A     2019-08-09  3              18440  
           B     2019-08-09  2              12840  
2038680547 A     2019-08-04  2              4070   
                 2019-08-17  1              3470   
                 2019-08-22  1              760    
                 2019-08-25  1              270    
           B     2019-08-04  1              2800   
                 2019-08-15  1              1680   
                 2019-08-29  1              2420   
2378935119 A     2019-08-10  1              58220  
                 2019-08-15  1              1650   
                 2019-08-17  1              310    
                 2019-08-22  1              3390   
           B     2019-08-09  1              30649  
                 2019-08-10  1              2640   
                 2019-08-25  1              21220  
                 2019-08-28  2              24860  
2458001652 A     2019-08-04  1              340    
                 2019-08-06  1              1620   
                 2019-08-07  1              17379  
                 2019-08-11  1              8280   
                 2019-08-22  1              5190   
           B     2019-08-05  1              2420   
                 2019-08-06  2              14030  
                 2019-08-08  1              5227   
                 2019-08-09  1              5227   
                 2019-08-25  1              2385   
4256040402 A     2019-08-01  1              5800   
                 2019-08-12  1              50     
                 2019-08-17  1              1550   
                 2019-08-20  1              17650  
                 2019-08-23  1              16950  
                 2019-08-27  1              1630   
                 2019-08-31  1              5170   
           B     2019-08-08  1              58220  
                 2019-08-16  1              56650  
                 2019-08-18  1              10970  
                 2019-08-31  1              1850

- Пять пользователей оказались в двух группах одновременно, они совершили достаточно много платежей и принесли значимую суму выручки, поэтому нужно очистить наши таблицы от этих пользователей, для корректного А/В теста

Отфильтруем таблицу visitors

In [12]:
inappropriate_orders_grouped = inappropriate_orders_grouped.reset_index()
inappropriate_orders_grouped['count'] = -1
inappropriate_orders_grouped = inappropriate_orders_grouped[['group', 'date', 'count']]
visitors_filt = visitors.merge(inappropriate_orders_grouped, on=['date', 'group'], how='left').fillna(0)
visitors_filt['visitors_'] = visitors_filt['visitors'] + visitors_filt['count']
visitors_filt = visitors_filt.drop(['visitors', 'count'], axis=1)
visitors_filt.columns = ['date', 'group', 'visitors']
visitors_filt['visitors'] = visitors_filt['visitors'].astype(int)

In [13]:
visitors

date group  visitors
0   2019-08-01  A     719     
1   2019-08-02  A     619     
2   2019-08-03  A     507     
3   2019-08-04  A     717     
4   2019-08-05  A     756     
..         ... ..     ...     
57  2019-08-27  B     720     
58  2019-08-28  B     654     
59  2019-08-29  B     531     
60  2019-08-30  B     490     
61  2019-08-31  B     718     

[62 rows x 3 columns]

- Так как в рамках этого A/B теста не пользователи, а сессии разделяли на группы, то проблемы в том, что пользователь попал в обе группы, нет. Он совершил несколько сессий, которые определились в разные группы.

### График кумулятивной выручки по группам.  

Строим график кумулятивной выручки по группам.

In [14]:
cum_revenue = orders_filt.groupby(['group', 'date'])['revenue'].sum().groupby(level=[0]).cumsum().to_frame()
cum_revenue = cum_revenue.reset_index()
cum_revenue['date'] = cum_revenue['date'].map(lambda x: dt.datetime.strptime(x, '%Y-%m-%d'))
cum_revenue_A = cum_revenue[cum_revenue['group'] == 'A'][['date', 'revenue']]
cum_revenue_B = cum_revenue[cum_revenue['group'] == 'B'][['date', 'revenue']]
cum_revenue

group       date  revenue
0   A    2019-08-01  142779 
1   A    2019-08-02  234381 
2   A    2019-08-03  346854 
3   A    2019-08-04  388030 
4   A    2019-08-05  474413 
.. ..           ...     ... 
57  B    2019-08-27  4579312
58  B    2019-08-28  4640510
59  B    2019-08-29  4746610
60  B    2019-08-30  4955833
61  B    2019-08-31  5068972

[62 rows x 3 columns]

In [15]:
trace_A = go.Scatter(
    x = cum_revenue_A['date'],
    y = cum_revenue_A['revenue'],
    mode = 'lines+markers',
    name = 'A'
)
trace_B = go.Scatter(
    x = cum_revenue_B['date'],
    y = cum_revenue_B['revenue'],
    mode = 'lines+markers',
    name = 'B'
)

layout = go.Layout(
    title='Кумулятивная выручка по группам',
)   

fig = go.Figure(data = [trace_A, trace_B], layout = layout)
iplot(fig)


- По графикам наблюдается увеличение кумулятивной выручки со временем.  
- 19 августа в группе В зафиксировано резкое увеличение, с дальнейшим сохранением прежней динамики.  
- Группа **В** лидирует.

### График кумулятивного среднего чека по группам   

Строим график кумулятивного среднего чека по группам.

In [16]:
cum_mean_check = orders_filt.groupby(['group','date'])['revenue'].mean().groupby(level=[0]).cumsum().to_frame().reset_index().round(1)
cum_mean_check
cum_mean_check_A = cum_mean_check[cum_mean_check['group']=='A']
cum_mean_check_B = cum_mean_check[cum_mean_check['group']=='B']

In [17]:
trace_A = go.Scatter(
    x = cum_mean_check_A['date'],
    y = cum_mean_check_A['revenue'],
    mode = 'lines+markers',
    name = 'A'
)
trace_B = go.Scatter(
    x = cum_mean_check_B['date'],
    y = cum_mean_check_B['revenue'],
    mode = 'lines+markers',
    name = 'B'
)

layout = go.Layout(
    title='Кумулятивный средний чек по группам',
    xaxis_title='Дата',
    yaxis_title='Средний чек'
)   

fig = go.Figure(data = [trace_A, trace_B], layout = layout)
iplot(fig)

- По графикам наблюдается увеличение кумулятивного среднего чека со временем.  
- До 19 августа группы меняли лидерство, находясь в тесной динамике.
- 19 августа в группе В зафиксировано резкое увеличение, с дальнейшим сохранением прежней динамики.  
- Группа **В** лидирует.

### График относительного изменения кумулятивного среднего чека группы B к группе A  

Строим график относительного изменения кумулятивного среднего чека группы B к группе A. 

In [18]:
mean_check_A = cum_mean_check.query('group == "A"')['revenue'].tolist()
mean_check_B = cum_mean_check.query('group == "B"')['revenue'].tolist()
mean_check_rel = [y/x - 1 for x,y in zip(mean_check_A, mean_check_B)]

cum_mean_check_rel = pd.DataFrame({'date': cum_mean_check['date'].unique(),
                                   'mean_check_rel': mean_check_rel})

In [19]:
layout = go.Layout(title='График относительного изменения кумулятивного среднего чека группы B к группе A',
                   xaxis_title='Дата',
                   yaxis_title='Средний чек')
fig = go.Figure(layout=layout)
fig.add_trace(go.Scatter(x=cum_mean_check_rel['date'],
                                      y=cum_mean_check_rel['mean_check_rel'],
                                      mode='lines+markers', line = dict(color='black')
                                     ))

fig.add_trace( go.Scatter(x = [min(cum_mean_check_rel['date']), max(cum_mean_check_rel['date'])], y=[0, 0], mode="lines",
                        line = dict(color='royalblue', width=4, dash='dash')))

fig.show()

- По графику можно наблюдать переходящий успех групп В и А. В конечном итоге за счет скачка 19 августа группа В лидирует.

### График кумулятивной конверсии по группам  

Строим график кумулятивной конверсии по группам.

In [20]:
orders_conversion = orders_filt.groupby(['date', 'group'])['transactionId'].nunique().to_frame().reset_index()
df_conversion = orders_conversion.merge(visitors_filt, on=['date', 'group'])
df_conversion['conversion'] = (df_conversion['transactionId'] / df_conversion['visitors']).round(4)
df_conversion = df_conversion[['date', 'group', 'conversion']]
df_conversion = (df_conversion
                          .groupby(['group', 'date'])['conversion']
                          .mean()
                          .groupby(level=[0])
                          .cumsum()
                          .round(2)
                          .to_frame()
                 )
df_conversion = df_conversion.reset_index()
df_conversion['date'] = df_conversion['date'].map(lambda x: dt.datetime.strptime(x, '%Y-%m-%d'))

In [21]:
df_conversion_A = df_conversion[df_conversion['group'] == 'A']
df_conversion_B = df_conversion[df_conversion['group'] == 'B']
layout = go.Layout(title='График кумулятивной конверсии по группам',
                  xaxis_title = 'Дата',
                  yaxis_title = 'Конверсия')

fig = go.Figure(layout=layout)
fig.add_trace(go.Scatter(x = df_conversion_A['date'],
                         y = df_conversion_A['conversion'], mode = 'lines+markers', name='A'))
fig.add_trace(go.Scatter(x = df_conversion_B['date'],
                         y = df_conversion_B['conversion'], mode = 'lines+markers', name='B'))
fig.show()

- По графику накопительная конверсия по группе B практически на всем протяжении лежит значительно выше накопительной конверсии группы A.

### График относительного изменения кумулятивной конверсии группы B к группе A  

Строим график относительного изменения кумулятивной конверсии группы B к группе A

In [22]:
rel_conv = df_conversion
rel_conv_A = rel_conv.query('group == "A"')['conversion'].tolist()
rel_conv_B = rel_conv.query('group == "B"')['conversion'].tolist()
mean_conv_rel = [y/x - 1 for x,y in zip(rel_conv_A, rel_conv_B)]

cum_mean_conv_rel = pd.DataFrame({'date': rel_conv['date'].unique(),
                                  'conversion': mean_conv_rel})
cum_mean_conv_rel
layout = go.Layout(title = 'График относительного изменения кумулятивной конверсии группы B к группе A',
                  xaxis_title = 'Дата',
                  yaxis_title = 'Конверсия')
fig = go.Figure(layout=layout)
fig.add_trace(go.Scatter(x = cum_mean_conv_rel['date'],
                         y = cum_mean_conv_rel['conversion'],
                         mode='lines+markers', line = dict(color='black')))

fig.add_trace( go.Scatter(x = [min(cum_mean_check_rel['date']), max(cum_mean_check_rel['date'])], y=[0, 0], mode="lines",
                        line = dict(color='royalblue', width=4, dash='dash')))

fig.add_trace( go.Scatter(x = [min(cum_mean_check_rel['date']), max(cum_mean_check_rel['date'])], y=[0.1, 0.1], mode="lines",
                        line = dict(color='grey', width=4, dash='dash')))

fig.show()

- При незначительном снижении группа В начиная с 3 августа наблюдается положительное изменение, 7 августа пересекает нулевую отметку и с 10 августа стабилизируется в небольшой коридор по показаниям превышающим 10%. 

### Точечный график количества заказов по пользователям  

Строим точечный график количества заказов по пользователям.

In [23]:
orders_by_users = orders_filt.groupby('visitorId', as_index=False).agg({'transactionId':'nunique'})
orders_by_users['count'] = range(1, len(orders_by_users)+1)

In [24]:
layout = go.Layout(title = 'Точечный график количества заказов по пользователям',
                  xaxis_title = 'Пользователи',
                  yaxis_title = 'Кол-во заказов')
fig = go.Figure(layout=layout)
fig.add_trace(go.Scatter(x=orders_by_users['count'], y=orders_by_users['transactionId'], mode='markers'))
fig.show()

- По графику видно, что наибольшее количество пользователей делают по одному заказу.  
- Однако присутствует значительная доля пользователей которые заделали по два заказа.  
- Незначительная доля по 3 заказов.  

### 95-й и 99-й перцентили количества заказов на пользователя  

Посчитаем 95-й и 99-й перцентили количества заказов на пользователя.  
Выберем границу для определения аномальных пользователей.

In [25]:
percentile_orders = np.percentile(orders_by_users['transactionId'], [95, 97.5, 99])
print(percentile_orders)

[1. 2. 2.]


- 99 перцентиль равен 4 заказам, не более 1% пользователей,   
- 97.5 перцентиль равен 2, не более 2,5% - **граница для определения аномальных пользователей**,   
- 95 перцентиль равен 2, не более 5% пользователей совершили по два заказа.

### Точечный график стоимостей заказов  

Строим точечный график стоимостей заказов.

In [26]:
revenue_by_users = orders_filt
revenue_by_users['count'] = range(1, len(revenue_by_users)+1)

In [27]:
layout = go.Layout(title = 'Точечный график стоимостей заказов',
                  xaxis_title = 'Пользователи',
                  yaxis_title = 'Стоимость')
fig = go.Figure(layout=layout)
fig.add_trace(go.Scatter(x=revenue_by_users['count'], y=revenue_by_users['revenue'], mode='markers'))
fig.show()

- Наблюдутся выброс на одного пользователя со стоимость заказа в 1.29 млн, построим нижем график с ограничим по У-оси.

In [28]:
layout = go.Layout(title = 'Точечный график стоимостей заказов до 205000',
                  xaxis_title = 'Пользователи',
                  yaxis_title = 'Стоимость', yaxis=dict(range=[0,205000]))
fig = go.Figure(layout=layout)
fig.add_trace(go.Scatter(x=revenue_by_users['count'], y=revenue_by_users['revenue'], mode='markers'))

fig.show()

- По графику видно, восполним заказы в диапазоне до 11-12 тыс.

### 95-й и 99-й перцентили стоимости заказов  

Посчитаем 95-й и 99-й перцентили стоимости заказов.  
Выберем границу для определения аномальных заказов.

In [29]:
percentile_revenue = np.percentile(revenue_by_users['revenue'], [95, 97.5, 99])
print(percentile_revenue)

[26785. 37400. 53904.]


- 99 перцентиль равен 53904, не более 1% пользователей,   
- 97.5 перцентиль равен 37400, не более 2,5% - **граница для определения аномальных заказов**,   
- 95 перцентиль равен 26785, не более 5% пользователей превышает сумма заказа.

**Сформируем очищенный датафрейм.**

In [30]:
clean_count_orders = orders_filt.groupby('visitorId')['transactionId'].count()
clean_count_orders = clean_count_orders[clean_count_orders > 2].index.tolist()

In [31]:
clean_sum_orders = orders_filt.groupby('visitorId')['revenue'].max()
clean_sum_orders = clean_sum_orders[clean_sum_orders > 37400].index.tolist()

In [32]:
orders_common = list(set(clean_count_orders + clean_sum_orders))

In [33]:
orders_final = orders_filt.query('visitorId not in @ orders_common')

### Статистическую значимость различий в конверсии между группами по «сырым» данным  

Посчитаем статистическую значимость различий в конверсии между группами по «сырым» данным.  
H0: конверсии в группах А и В по «сырым» данным не отличаются/ равны.  
H1: конверсии в группах А и В по «сырым» данным отличаются/ не равны.  
Уровень значимости :0.05

In [34]:
#p-value  по сырым данным
orders_conversion = orders_filt.groupby(['date', 'group'])['visitorId'].nunique().to_frame().reset_index()

df_conv_comparison = orders_conversion.merge(visitors, on=['date', 'group'])
df_conv_comparison['conversion'] = (df_conv_comparison['visitorId'] / df_conv_comparison['visitors']).round(4)
df_conv_comparison = df_conv_comparison[['date', 'group', 'conversion']]

conv_A = df_conv_comparison.query('group == "A"')['conversion'].tolist()
conv_B = df_conv_comparison.query('group == "B"')['conversion'].tolist()
conv_A_mean = sum(conv_A) / len(conv_A)
conv_B_mean = sum(conv_B) / len(conv_B)

alpha = .05
p_val_conv = stats.mannwhitneyu(conv_A, conv_B)[1].round(5)

print('P-value по конверсии до очистки данных: {}'.format(p_val_conv))
if (p_val_conv < alpha):
    print("Отвергаем нулевую гипотезу")
else:
    print("Не получилось отвергнуть нулевую гипотезу")
print('Отношения средних B/A по конверсии (до очистки данных): {:.2f}'.format(conv_B_mean/conv_A_mean))

P-value по конверсии до очистки данных: 0.03653
Отвергаем нулевую гипотезу
Отношения средних B/A по конверсии (до очистки данных): 1.19


### Cтатистическая значимость различий в среднем чеке заказа между группами по «сырым» данным  

Посчитаем статистическую значимость различий в среднем чеке заказа между группами по «сырым» данным.   
H0: среднем чеке заказа в группах А и В по «сырым» данным не отличаются/ равны.  
H1: среднем чеке заказа в группах А и В по «сырым» данным отличаются/ не равны.
Уровень значимости :0.05

In [35]:
mean_check_A = orders_filt[orders_filt['group'] == 'A']['revenue']
mean_check_B = orders_filt[orders_filt['group'] == 'B']['revenue']
mean_check_A_final = sum(mean_check_A) / len(mean_check_A)
mean_check_B_final = sum(mean_check_B) / len(mean_check_B)

p_val_mean_check = stats.mannwhitneyu(mean_check_A, mean_check_B)[1].round(2)

alpha = .05

print('P-value по конверсии до очистки данных: {}'.format(p_val_mean_check))
if (p_val_mean_check < alpha):
    print("Отвергаем нулевую гипотезу")
else:
    print("Не получилось отвергнуть нулевую гипотезу")
    
#print('Отношения B/A в среднем чеке заказа (до очистки данных): {:.2f}'.format(mean_check_B_final/mean_check_A_final))

print('Отношения B/A в среднем чеке заказа (до очистки данных): {0:.3f}'.format(orders[orders['group']=='B']['revenue'].mean()/orders[orders['group']=='A']['revenue'].mean()-1)) 

P-value по конверсии до очистки данных: 0.83
Не получилось отвергнуть нулевую гипотезу
Отношения B/A в среднем чеке заказа (до очистки данных): 0.259


In [36]:
#КОД
print('{0:.3f}'.format(orders[orders['group']=='B']['revenue'].mean()/orders[orders['group']=='A']['revenue'].mean()-1)) 

0.259


### Статистическую значимость различий в конверсии между группами по «очищенным» данным  

Посчитаем статистическую значимость различий в конверсии между группами по «очищенным» данным.  
H0: конверсии в группах А и В по «очищенным» данным не отличаются/ равны.  
H1: конверсии в группах А и В по «очищенным» данным отличаются/ не равны.  
Уровень значимости :0.05

In [37]:
#p-value - по очищенным данным
orders_conversion_clean = orders_final.groupby(['date', 'group'])['visitorId'].nunique().to_frame().reset_index()

df_conv_comparison_clean = orders_conversion_clean.merge(visitors, on=['date', 'group'])
df_conv_comparison_clean['conversion'] = (df_conv_comparison_clean['visitorId']/df_conv_comparison_clean['visitors'])
df_conv_comparison_clean['conversion'] = df_conv_comparison_clean['conversion'].round(4)
df_conv_comparison_clean = df_conv_comparison_clean[['date', 'group', 'conversion']]

conv_clean_A = df_conv_comparison_clean.query('group == "A"')['conversion'].tolist()
conv_clean_B = df_conv_comparison_clean.query('group == "B"')['conversion'].tolist()
conv_A_mean_cleared = sum(conv_clean_A)/len(conv_clean_A)
conv_B_mean_cleared = sum(conv_clean_B)/len(conv_clean_B)

alpha = .05

p_val_conv_clean = stats.mannwhitneyu(conv_clean_A, conv_clean_B)[1].round(5)

print('P-value по конверсии после очистки данных: {}'.format(p_val_conv_clean))
if (p_val_conv_clean < alpha):
    print("Отвергаем нулевую гипотезу")
else:
    print("Не получилось отвергнуть нулевую гипотезу")

print('Отношения средних B/A по конверсии (после очистки данных): {:.2f}'.format
                                                                          (conv_B_mean_cleared/conv_A_mean_cleared))

P-value по конверсии после очистки данных: 0.01943
Отвергаем нулевую гипотезу
Отношения средних B/A по конверсии (после очистки данных): 1.21


### Статистическую значимость различий в среднем чеке заказа между группами по «очищенным» данным  

Посчитаем статистическую значимость различий в среднем чеке заказа между группами по «очищенным» данным.  
H0: среднем чеке заказа в группах А и В по «очищенным» данным не отличаются/ равны.  
H1: среднем чеке заказа в группах А и В по «очищенным» данным отличаются/ не равны.
Уровень значимости :0.05

In [38]:
mean_check_clean_A = orders_final[orders_final['group'] == 'A']['revenue']
mean_check_clean_B = orders_final[orders_final['group'] == 'B']['revenue']
mean_check_clean_A_final = sum(mean_check_clean_A) / len(mean_check_clean_A)
mean_check_clean_B_final = sum(mean_check_clean_B) / len(mean_check_clean_B)

p_val_mean_check_clean = stats.mannwhitneyu(mean_check_clean_A, mean_check_clean_B)[1].round(2)

alpha = .05

print('P-value по конверсии после очистки данных: {}'.format(p_val_mean_check_clean))
if (p_val_mean_check_clean < alpha):
    print("Отвергаем нулевую гипотезу")
else:
    print("Не получилось отвергнуть нулевую гипотезу")
    
print('Отношения B/A в среднем чеке заказа (после очистки данных): {:.2f}'.format
                                                                          (mean_check_clean_B_final/mean_check_clean_A_final))

P-value по конверсии после очистки данных: 0.94
Не получилось отвергнуть нулевую гипотезу
Отношения B/A в среднем чеке заказа (после очистки данных): 1.02


При сравнении двух гипотез использовался метод равенства средних двух независимых совокупностей,
который дает очевидную наглядность равенства или нет, отвержения нулевой гипотезы или нет.
Критерием здесь служит p-значение, принимаемое на усмотрение аналитика. В данном случае принят показатель 0.05 (5%).   

**Результаты А/В теста:**  

- В конверсии между группами А и В есть наблюдаемые различия, есть статистической значимости различий по сырым и очищенным данным.

P-value по конверсии до очистки данных: 0.03653  
Отвергаем нулевую гипотезу  
Отношения средних B/A по конверсии (до очистки данных): 1.19  

P-value по конверсии после очистки данных: 0.01943  
Отвергаем нулевую гипотезу  
Отношения средних B/A по конверсии (после очистки данных): 1.21

**После очистки данных отношение между группами увеличилось на 2% в пользу группы В.**

- По средним чекам между группами А и В нет наблюдаемые различия, нет статистическая значимость различий как по сырым данным, так и по очищенным.  

P-value по конверсии до очистки данных: 0.83  
Не получилось отвергнуть нулевую гипотезу  
Отношения B/A в среднем чеке заказа (до очистки данных): 0.259 

P-value по конверсии после очистки данных: 0.94  
Не получилось отвергнуть нулевую гипотезу  
Отношения B/A в среднем чеке заказа (после очистки данных): 1.02 

**После очистки данных отношение B/A в среднем чеке заказа уменьшилось на 27% в пользу группы А.**

### Общий вывод   

Проведенный тест изменился только на половину - по конверсии. Но это тоже хорошо (хотя, надо принимать во внимание, что было важнее - увеличить средний чек или конверсию). А т.к. стат.значимой разницы по среднему чеку нет (чек не увеличился, но и не уменьшился) то при выросшей конверсии с сохранившимся средним чеком выручка будет так-же расти! И значит мы заканчиваем анализ A/B-теста - он оказался успешным!

**На основании результатов проведенного теста принято решение:**  
- Остановить тест, зафиксировать победу **В** группы.
- Конверсия у группы **В** выше чем у группы **А** по сырым и очищенным данным.  
- Конверсия по среднему чеку не имеет значимых различий между группами.